lib/Imagine/Image/AbstractImage.php

```diff
@@ -63,33 +68,60 @@ public function thumbnail(BoxInterface $size, $mode = ImageInterface::THUMBNAIL_
             $ratio = max($ratios);
         }
 
-        if ($mode === ImageInterface::THUMBNAIL_OUTBOUND) {
+        if ($mode === ImageInterface::THUMBNAIL_INSET) {
+            $imageSize = $imageSize->scale($ratio);
+            $thumbnail->resize($imageSize, $filter);
+        } else {
             if (!$imageSize->contains($size)) {
+                if ($allowUpscale) {
+                    $imageSize = $imageSize->scale($ratio);
+                    $thumbnail->resize($imageSize, $filter);
+                }
                 $size = new Box(
                     min($imageSize->getWidth(), $size->getWidth()),
                     min($imageSize->getHeight(), $size->getHeight())
                 );
             } else {
-                $imageSize = $thumbnail->getSize()->scale($ratio);
+                $imageSize = $imageSize->scale($ratio);
                 $thumbnail->resize($imageSize, $filter);
             }
             $thumbnail->crop(new Point(
                 max(0, round(($imageSize->getWidth() - $size->getWidth()) / 2)),
                 max(0, round(($imageSize->getHeight() - $size->getHeight()) / 2))
             ), $size);
-        } else {
-            if (!$imageSize->contains($size)) {
-                $imageSize = $imageSize->scale($ratio);
-                $thumbnail->resize($imageSize, $filter);
-            } else {
-                $imageSize = $thumbnail->getSize()->scale($ratio);
-                $thumbnail->resize($imageSize, $filter);
-            }
         }
 
         return $thumbnail;
     }
 
+    /**
+     * Check the settings argument in thumbnail() method.
+     *
+     * @param int $settings
+     */
+    private function checkThumbnailSettings($settings)
+    {
+        // Preserve BC until version 1.0
+        if ($settings === 'inset') {
+            $settings = ImageInterface::THUMBNAIL_INSET;
+        } elseif ($settings === 'outbound') {
+            $settings = ImageInterface::THUMBNAIL_OUTBOUND;
+        }
+
+        $allSettings = ImageInterface::THUMBNAIL_INSET | ImageInterface::THUMBNAIL_OUTBOUND | ImageInterface::THUMBNAIL_UPSCALE;
+
+        if (!is_int($settings) || ($settings & ~$allSettings)) {
+            throw new InvalidArgumentException('Invalid setting specified');
+        }
+
+        if ($settings & ImageInterface::THUMBNAIL_INSET &&
+            $settings & ImageInterface::THUMBNAIL_OUTBOUND) {
+            throw new InvalidArgumentException('Only one mode should be specified');
```

Comment:
Michele, see also here. By using "settings", we can distinguish "all settings" / "only modes".

Reply:
I simply think that the word *mode* gives much more relevance to the argument than *setting*. Setting seems like an optional stuff, mode highlights its relevance. BTW I'm not going to waste time on this subject :wink: